api_test/src/test/java/ecse429/group7/performance/TestProjectPerformance.java

```diff
@@ -0,0 +1,214 @@
+package ecse429.group7.performance;
+
+import org.junit.Test;
+
+import kong.unirest.Unirest;
+import kong.unirest.HttpResponse;
+import kong.unirest.JsonNode;
+
+public class TestProjectPerformance extends BasePerformanceTest {
+    
+    public void addRandomProject()
+    {
+        HttpResponse<JsonNode> response = Unirest.post("/projects")
+                .body("{\n\"description\":\"" + getRandomString() + "\",\n    \"active\":" + getRandomBool()
+                        + ",\n    \"completed\":" + getRandomBool() + ",\n    \"title\":\"" + getRandomString()
+                        + "\"\n}")
+                .asJson();
+
+        project_id_list.add(response.getBody().getObject().getInt("id"));
+    }
+
+    public void addManyProjects(int number_projects)
+    {
+        for (int i = 0; i < number_projects; i++) {
+            addRandomProject();
+        }
+    }
+
+    public void removeLastProject()
+    {
+        Unirest.delete("/projects/" + project_id_list.getLast()).header("Content-Type", "application/json").asJson();
+        project_id_list.removeLast();
+    }
+
+    public void changeLastProject()
+    {
+        Unirest.put("/projects/" + project_id_list.getLast())
+        .body("{\n\"description\":\"" + getRandomString() + "\",\n    \"active\":" + getRandomBool()
+                + ",\n    \"completed\":" + getRandomBool() + ",\n    \"title\":\"" + getRandomString()
+                + "\"\n}")
+        .asJson();
+    }
+
+    public void testManyProjectsAddTime(int number_projects)
+    {
+        long start_whole = System.nanoTime();
+
+        // Setup
+        addManyProjects(number_projects);
+
+        // Add one project
+        long start_single = System.nanoTime();
+        addRandomProject();
+        long finish_single = System.nanoTime();
+
+        // Reset state
+        resetState();
+
+        long finish_whole = System.nanoTime();
+        System.out.println("Test Add Project with " + number_projects + " Projects in Server:");
+        System.out.println("\tTotal Test Time: " + (finish_whole - start_whole) + " ns");
+        System.out.println("\tSingle Project Add Time: " + (finish_single - start_single) + " ns");
+    }
+
+    public void testManyProjectDeleteTime(int number_projects)
```

Comment:
it's a huge bummer that our java version doesn't have my sweet sweet lambda expressions, cause a lot of these timed methods are very similar to each other and could likely be refactored to something more general that takes some lambdas to run and time. I don't think it's worth spending time on this, I just think if i'm giving my full takes on this code that's something i'd thinking about fixing in an ideal world where we weren't all overwhelmed with work for other classes lol

Reply:
Yeah the best way to do this would have been to create an interface and have objects implement the interface but honestly, for what we want to do, the overhead would be a bit ridiculous in terms of the number of classes